### PR TITLE
KP-506 [PagerDuty Create Incident] add Parameters

### PR DESCRIPTION
**Pull request by xkcm**
* Add new parameters to **Create Incident** method
* Code refactoring

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+node_modules/
+yarn.lock
```
Comment (arispen) [inline .gitignore]: please provide the .gitignore from the plugin template
https://github.com/Kaholo/kaholo-plugin-template

File: app.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+} = require("./helpers");
+
+/**
+ * Creates a new Incident
+ * Based on Docs here: https://developer.pagerduty.com/api-reference/reference/REST/openapiv3.json/paths/~1incidents/post
+ */
+async function createNewIncident({ params }, settings) {
+  const {
+    SERVICE_ID, ASSIGNEE, TITLE, EMAIL, TOKEN, DETAILS, INCIDENT_KEY,
+    URGENCY, PRIORITY, ESCALATION_POLICY, CONFERENCE_NUMBER, CONFERENCE_URL,
```
Comment (arispen) [inline app.js]: are we going to parse it in any way?

File: app.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  return result;
+}
+
+/**
+ * Send a change events to PD Events API.
+ * Based on the docs in this path: https://developer.pagerduty.com/api-reference/reference/events-v2/openapiv3.json/paths/~1change~1enqueue/post
+ */
+async function createChangeEvent({ params }, settings) {
+  const {
+    SUMMARY, ROUTING_KEY, SOURCE, CUSTOM_DETAILS,
```
Comment (arispen) [inline app.js]: parsing?

File: autocomplete.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  performApiRequest, constructAuthorizationHeader, mapSettingsAndParams,
+  filterAutocompleteOptions, mapAutocompleteOptions,
+} = require("./helpers");
+
+const AUTOCOMPLETE_OPTIONS_LIMIT = 5;
+
+function createAutocompleteFunction({ path, resultArrayPath, requestParams }) {
+  return async (query, pluginSettings, actionParams) => {
+    const { settings, params } = mapSettingsAndParams(pluginSettings, actionParams);
+    const result = await performApiRequest({
```
Comment (arispen) [inline autocomplete.js:17]: error handling?

File: config.json [hunk 5]
```diff
@@ -0,0 +1,173 @@
+{
+  "name": "pager-duty",
+  "viewName": "PagerDuty",
+  "type": "executer",
+  "execProgram": "node",
+  "main": "app.js",
+  "exposeRoute": true,
+  "imgUrl": "pg.png",
+  "version": "1.1.1",
```
Comment (arispen) [inline config.json]: I think you can increase the version to 2.0.0

File: package.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "kaholo-plugin-pagerduty",
+  "description": "Kaholo Plugin for Datadog",
```
Comment (arispen) [inline package.json]: I think you can remove it

File: config.json [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  "description": "Integration with PagerDuty REST API integration",
```
Comment (arispen) [inline config.json]: integration with integration?

File: config.json [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+          "description": "The email address of a valid user associated with the account making the request.",
+          "type": "string",
+          "learnUrl": "https://developer.pagerduty.com/api-reference/b3A6Mjc0ODE0MA-create-an-incident"
+        },
+        {
+          "name": "TITLE",
+          "viewName": "Title",
+          "description": "A succinct description of the nature, symptoms, cause, or effect of the incident.",
+          "type": "string",
+          "required": true,
```
Comment (arispen) [inline config.json]: I think we should put required params first

File: config.json [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+          "type": "vault",
+          "learnUrl": "https://developer.pagerduty.com/api-reference/b3A6Mjc0ODE0MA-create-an-incident"
+        },
+        {
+          "name": "EMAIL",
+          "viewName": "Email",
+          "description": "The email address of a valid user associated with the account making the request.",
+          "type": "string",
+          "learnUrl": "https://developer.pagerduty.com/api-reference/b3A6Mjc0ODE0MA-create-an-incident"
+        },
```
Comment (arispen) [inline config.json]: these are not required?

File: helpers.js [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+ * @param {Object} responseData
+ */
+function parsePagerDutyErrorMessage(responseData) {
+  const { message, errors } = (responseData || {}).error || {};
+  let errorResponse = `${message}\n`;
+  if (errors) {
+    errorResponse += errors.join("\n");
+  }
+  if (!message) {
+    errorResponse = JSON.stringify(responseData);
```
Comment (arispen) [inline helpers.js]: add error handling for stringify

File: helpers.js [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+ * Sends the request to the PagerDuty API endpoint
+ * @param {{
+ *  method: "GET" | "POST" | "PUT" | "DELETE" | "PATCH"
+ *  body?: Object
+ *  path?: string
+ *  params?: Object
+ *  headers?: Object
+ * }} options
+ * @returns {Object}
+ */
```
Comment (arispen) [inline helpers.js]: I think we can remove jsdocs

File: helpers.js [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    }
+    throw new Error(errorMessage);
+  }
+}
+
+/**
+ * Removes keys from object if the value is undefined or null
+ * @param {Object} o
+ */
+function tidyObject(o) {
```
Comment (arispen) [inline helpers.js]: are we sure that there is no better way to handle that using lodash?

**Review by arispen — CHANGES_REQUESTED** [review]

File: config.json (hunk 9 above)
Comment (xkcm) [reply config.json]: The Token can be provided in the Plugin's settings so it should not be required here, but the Email should be required, thanks for pointing this out.

**Review by arispen — DISMISSED** [review]

**Review by arispen — APPROVED** [review]

**Pull request merged**